{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Author:** Dan Shea  \n",
    "**Date:** 2019.11.15  \n",
    "**Description:**  \n",
    "#### Best Reciprocal `blastp` Hit\n",
    "Given amino acid sequences for four _Brassicaceae_ organisms (2 files for _B. rapa_ --version 1.5 & version 3), we will run pairwise `blastp` and report the best hit for each of the 20 possible pairings (20 and not 10 because here order is important). The results will then be merged for pairings that show reciprocity. Those genes that do not show reciprocity for their blast results will be saved in another file. We will make use of `Bio.Blast.Applications` which can run command line blastp from within a python program using `NcbiblastxCommandline`.\n",
    "\n",
    "__UPDATE__ - It turns out that that multithreading is not being used when blastp is invoked via `NcbiblastxCommandline`. So I have a script called `run_blast.sh` (what else! ;-) This script fires off blastp sequentially and asks for 10 threads. We will import the XML output in this workbook to determine the reciprocal best hits."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIXML\n",
    "import os\n",
    "import os.path\n",
    "import itertools\n",
    "import operator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# See UPDATE\n",
    "# def run_blastp(queryfile, subjectfile):\n",
    "#     query = os.path.basename(queryfile).split('.faa')[0]\n",
    "#     subject = os.path.basename(subjectfile).split('.faa')[0]\n",
    "#     outfile = '_vs_'.join([query, subject]) + '.xml'\n",
    "#     # Construct the command line to run such that we only return the best hit (max_target_seqs=1)\n",
    "#     blastp_cline = NcbiblastxCommandline(cmd='blastp', out=outfile, outfmt=5, query=queryfile, subject=subjectfile, max_target_seqs=1, num_threads=10)\n",
    "#     # invoke the blastp command\n",
    "#     blastp_cline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "inputfiles = ['A_thaliana.faa', 'B_napus.faa', 'B_oleracea.faa', 'B_rapa.faa', 'B_rapa_v1.5.faa']\n",
    "inputfiles = list(itertools.starmap(operator.add, itertools.product(['references/'],inputfiles)))\n",
    "idx = 0\n",
    "combinations = list()\n",
    "for ifile in inputfiles:\n",
    "    combinations.extend(list(itertools.product([ifile], inputfiles[0:idx]+inputfiles[idx+1:])))\n",
    "    idx += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# See UPDATE\n",
    "# for combination in combinations:\n",
    "#     run_blastp(*combination)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('references/A_thaliana.faa', 'references/B_napus.faa'),\n",
       " ('references/A_thaliana.faa', 'references/B_oleracea.faa'),\n",
       " ('references/A_thaliana.faa', 'references/B_rapa.faa'),\n",
       " ('references/A_thaliana.faa', 'references/B_rapa_v1.5.faa'),\n",
       " ('references/B_napus.faa', 'references/A_thaliana.faa'),\n",
       " ('references/B_napus.faa', 'references/B_oleracea.faa'),\n",
       " ('references/B_napus.faa', 'references/B_rapa.faa'),\n",
       " ('references/B_napus.faa', 'references/B_rapa_v1.5.faa'),\n",
       " ('references/B_oleracea.faa', 'references/A_thaliana.faa'),\n",
       " ('references/B_oleracea.faa', 'references/B_napus.faa'),\n",
       " ('references/B_oleracea.faa', 'references/B_rapa.faa'),\n",
       " ('references/B_oleracea.faa', 'references/B_rapa_v1.5.faa'),\n",
       " ('references/B_rapa.faa', 'references/A_thaliana.faa'),\n",
       " ('references/B_rapa.faa', 'references/B_napus.faa'),\n",
       " ('references/B_rapa.faa', 'references/B_oleracea.faa'),\n",
       " ('references/B_rapa.faa', 'references/B_rapa_v1.5.faa'),\n",
       " ('references/B_rapa_v1.5.faa', 'references/A_thaliana.faa'),\n",
       " ('references/B_rapa_v1.5.faa', 'references/B_napus.faa'),\n",
       " ('references/B_rapa_v1.5.faa', 'references/B_oleracea.faa'),\n",
       " ('references/B_rapa_v1.5.faa', 'references/B_rapa.faa')]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combinations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Alrighty then! From here, we will parse the XML output generated by `blastp` and construct our BRBHs for each pairing\n",
    "`blastp` produced `.xml` files as output (`-outfmt 5`). Now, we will make use of the blast output parser in `Bio.Blast.NCBIXML`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "blastp_files = ['A_thaliana_B_napus.xml','A_thaliana_B_oleracea.xml','A_thaliana_B_rapa.xml','A_thaliana_B_rapa_v1.5.xml',\n",
    "                'B_napus_A_thaliana.xml','B_napus_B_oleracea.xml','B_napus_B_rapa.xml','B_napus_B_rapa_v1.5.xml',\n",
    "                'B_oleracea_A_thaliana.xml','B_oleracea_B_napus.xml','B_oleracea_B_rapa.xml','B_oleracea_B_rapa_v1.5.xml',\n",
    "                'B_rapa_A_thaliana.xml','B_rapa_B_napus.xml','B_rapa_B_oleracea.xml','B_rapa_B_rapa_v1.5.xml',\n",
    "                'B_rapa_v1.5_A_thaliana.xml','B_rapa_v1.5_B_napus.xml','B_rapa_v1.5_B_oleracea.xml','B_rapa_v1.5_B_rapa.xml',]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_blast(filename):\n",
    "    results = list()\n",
    "    with open(filename, 'r') as fh:\n",
    "        blast_records = NCBIXML.parse(fh)\n",
    "        for blast_record in blast_records:\n",
    "            # Only use the gene names to ensure the later reciprocal lookup will work\n",
    "            # Found there are extra spaces in B. oleracea output that was causing lookup to fail\n",
    "            query = blast_record.query.split(' ')[0]\n",
    "            if len(blast_record.alignments) > 0:\n",
    "                hit = blast_record.alignments[0].hit_def.split(' ')[0]\n",
    "                results.append([query, hit])\n",
    "            else:\n",
    "                results.append([query, 'NA'])\n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's construct keys for the resultsa dictionary\n",
    "blast_result_keys = list()\n",
    "for a, b in combinations:\n",
    "    a = os.path.basename(a).split('.faa')[0]\n",
    "    b = os.path.basename(b).split('.faa')[0]\n",
    "    blast_result_keys.append(':'.join([a,b]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['A_thaliana:B_napus',\n",
       " 'A_thaliana:B_oleracea',\n",
       " 'A_thaliana:B_rapa',\n",
       " 'A_thaliana:B_rapa_v1.5',\n",
       " 'B_napus:A_thaliana',\n",
       " 'B_napus:B_oleracea',\n",
       " 'B_napus:B_rapa',\n",
       " 'B_napus:B_rapa_v1.5',\n",
       " 'B_oleracea:A_thaliana',\n",
       " 'B_oleracea:B_napus',\n",
       " 'B_oleracea:B_rapa',\n",
       " 'B_oleracea:B_rapa_v1.5',\n",
       " 'B_rapa:A_thaliana',\n",
       " 'B_rapa:B_napus',\n",
       " 'B_rapa:B_oleracea',\n",
       " 'B_rapa:B_rapa_v1.5',\n",
       " 'B_rapa_v1.5:A_thaliana',\n",
       " 'B_rapa_v1.5:B_napus',\n",
       " 'B_rapa_v1.5:B_oleracea',\n",
       " 'B_rapa_v1.5:B_rapa']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast_result_keys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now parse the results and store them in the dictionary\n",
    "blast_result_dict = dict()\n",
    "for key in blast_result_keys:\n",
    "    blastp_file = '_'.join(key.split(':')) + '.xml'\n",
    "    blast_result_dict[key] = parse_blast(blastp_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking A_thaliana:B_napus and B_napus:A_thaliana\n",
      "Checking A_thaliana:B_oleracea and B_oleracea:A_thaliana\n",
      "Checking A_thaliana:B_rapa and B_rapa:A_thaliana\n",
      "Checking A_thaliana:B_rapa_v1.5 and B_rapa_v1.5:A_thaliana\n",
      "Checking B_napus:B_oleracea and B_oleracea:B_napus\n",
      "Checking B_napus:B_rapa and B_rapa:B_napus\n",
      "Checking B_napus:B_rapa_v1.5 and B_rapa_v1.5:B_napus\n",
      "Checking B_oleracea:B_rapa and B_rapa:B_oleracea\n",
      "Checking B_oleracea:B_rapa_v1.5 and B_rapa_v1.5:B_oleracea\n",
      "Checking B_rapa:B_rapa_v1.5 and B_rapa_v1.5:B_rapa\n"
     ]
    }
   ],
   "source": [
    "reciprocal_hits = dict()\n",
    "checked = [0 for _ in range(len(blast_result_keys))]\n",
    "A_idx = 0\n",
    "for key in blast_result_keys:\n",
    "    # A is the current key\n",
    "    A = key\n",
    "    # B is the the corresponding reverse comparison key\n",
    "    B_idx = blast_result_keys.index(':'.join(reversed(key.split(':'))))\n",
    "    B = blast_result_keys[B_idx]\n",
    "    # If they have not alread been checked, check them otherwise move to the next pair\n",
    "    if (checked[A_idx] == 0) and (checked[B_idx] == 0):\n",
    "        # mark them as having been checked\n",
    "        checked[A_idx] = 1\n",
    "        checked[B_idx] = 1\n",
    "        print(f'Checking {A} and {B}')\n",
    "        # find the reciprocal best blast hits\n",
    "        for A_hit in blast_result_dict[A]:\n",
    "            # For each hit in A we try to lookup the reverse in B\n",
    "            # if we get a ValueError, that means that there is no reciprocal best hit\n",
    "            try:\n",
    "                B_lookup = list(reversed(A_hit))\n",
    "                B_hit = blast_result_dict[B].index(B_lookup)\n",
    "                if A in reciprocal_hits:\n",
    "                    reciprocal_hits[A].append(A_hit)\n",
    "                else:\n",
    "                    reciprocal_hits[A] = [A_hit]\n",
    "            except ValueError as e:\n",
    "                pass\n",
    "    else:\n",
    "        pass\n",
    "    A_idx += 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['A_thaliana:B_napus', 'A_thaliana:B_oleracea', 'A_thaliana:B_rapa', 'A_thaliana:B_rapa_v1.5', 'B_napus:B_oleracea', 'B_napus:B_rapa', 'B_napus:B_rapa_v1.5', 'B_oleracea:B_rapa', 'B_oleracea:B_rapa_v1.5', 'B_rapa:B_rapa_v1.5'])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reciprocal_hits.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write out results to file\n",
    "for k in reciprocal_hits:\n",
    "    outfile = k + '.brbh.tsv'\n",
    "    with open(outfile, 'w') as ofh:\n",
    "        for entry in reciprocal_hits[k]:\n",
    "            ofh.write('\\t'.join(entry))\n",
    "            ofh.write('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for k in blast_result_dict:\n",
    "#     outfile = k + '.blastp.tsv'\n",
    "#     with open(outfile, 'w') as ofh:\n",
    "#         for entry in blast_result_dict[k]:\n",
    "#             ofh.write('\\t'.join(entry))\n",
    "#             ofh.write('\\n')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
