{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Author: Dan Shea  \n",
    "Date: 2019.09.11  \n",
    "### Repulsion vs. Coupling of genotypes at loci showing interchromosomal LD\n",
    "We compared the set of SNPs common to all 20 founders (3,745 SNPs) by partitioning the genome (by chromosome) into $\\mathscr{K}$ partitions. In rice this yields $| \\mathscr{K} | = 12$. All of the SNPs $\\mathscr{s}_{i} \\in \\mathscr{k}_{1}$ were then compared to all SNPs $\\mathscr{s}_{j} \\in {\\mathscr{K} - \\mathscr{k}_{1}}$. This yielded all $\\langle \\mathscr{s}_{i}, \\mathscr{s}_{j} \\rangle$ pairs of SNPs. Since order is not important, we can proceed to the next chromosome $\\mathscr{k}_2$ and compare SNPs $\\mathscr{s}_{i}$ with $\\mathscr{s}_{j} \\in {\\mathscr{K} - \\mathscr{k}_{1} - \\mathscr{k}_{2}$. We have done this for the first 11 chromosomes, generating all possible SNP combinations between SNPs on different chromosomes.\n",
    "\n",
    "In the absence of _Linkage Disequilibrium_ (LD), we can expect a single SNP has $P(x) = 0.5$. Since we are looking at an _independent_ event at each position in our dimer (i.e. - the $\\langle \\mathscr{s}_{i}, \\mathscr{s}_{j} \\rangle$ pair) the _expected frequency of observation_ ($E(f_{obs})$) for any given dimer is simply $P(x,y) = P(x) \\cdot P(y) = 0.5 \\cdot 0.5 = 0.25$\n",
    "\n",
    "We have an expected $f_{obs}$ and our actual $f^{\\prime}_{obs}$, so we performed a $\\chi^{2}$ goodness-of-fit test to examine if the genotypes of the dimers observed was in line with the expectation that there is no LD between the two loci. (This expectation of no LD between loci on two different chromosomes being our $H_{0}$ for the $\\chi^{2}$ test.)\n",
    "\n",
    "Given that our SNP set is comprised of 3,745 loci and we want to examine pairs of SNPs, comparing every SNP to every other SNP would combinatorically be $C_{k}(n) = \\frac{n!}{k!\\cdot(n-k)!} = C_{2}(3,745) = \\frac{3,745 \\cdot 3,744}{2} = 7,010,640$. However, our actual number of comparisons was smaller than this, because we further restricted pairs to be on different chromosomes.\n",
    "\n",
    "After we completed our tests and receive p-values, we then performed an FDR correction on them using the `statsmodels.stats` `multitest.fdrcorrection()` function. The resulting FDR-corrected p-values (i.e. - q-values) were then compared to $\\alpha = 0.001$ and deemed significant if $q<\\alpha$.\n",
    "\n",
    "Inter-chromosomal Linkage Disequilibrium can provide evidence of epistatic interaction. _Epistasis_ is the effect of one gene's genotype influencing the expression of another gene. However, another interesting thing to examine is whether the genotypes of the two positions are in __repulsion__ or __coupling__. That is to say, is the tendency of the two loci to be homozygous for the same parent genotype (i.e. - in coupling) or are they heterozygous having the genotype of one parent at one SNP and the genotype of another parent at the other SNP (i.e. - in repulsion). This is important because it can reveal epistatic interactions between genes where offspring viability is determined by genotype.\n",
    "\n",
    "One example of this is seen in plant qualitative resistance to biotrophic pathogens. To defend themselves from pathogens, plants frequently feature a rapid hypersensitive response (HR). This is characterized by rapid and localized cell death (i.e. - apoptosis) to restrict the pathogen's replication during the early stages of the infection (Pontier et al., 1998). HR-associated cell death is a kind of programmed cell death (PCD), and plants with lethal combinations of R genes can trigger HR in the absence of a pathogen infection rendering the genotype non-viable.\n",
    "\n",
    "By checking to see if particular regions between chromosomes are in _coupling_ or _repulsion_, we can help identify potential areas of interest to examine in more detail to see if genes within the regions are exhibiting genotypic skew.\n",
    "\n",
    "行きましょう！"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from scipy import stats\n",
    "import numpy as np\n",
    "from statsmodels.stats import multitest\n",
    "import os\n",
    "import os.path\n",
    "from collections import OrderedDict, namedtuple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = ['N01','N03','N04','N05','N06','N07','N08','N09','N10','N11',\n",
    "           'N12','N13','N14','N16','N17','N18','N19','N20','N21','N22',]\n",
    "founders = ['KASALATH','KEIBOBA','SHONI','TUPA_121-3','SURJAMUKHI','RATUL','BADARI_DHAN','KALUHEENATI','JAGUARY','REXMONT',\n",
    "            'URASAN','TUPA_729','DEE_JIAO_HUA_LUO','NERICA_1','TAKANARI','C8005','MOUKOTOU','NORTAI','SESIA','HAYAYUKI',]\n",
    "datadirs = ['_'.join([x, y]) for x, y in zip(samples, founders)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/dshea/anaconda3/lib/python3.6/site-packages/numpy/lib/arraysetops.py:569: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  mask |= (ar1 == a)\n"
     ]
    }
   ],
   "source": [
    "# Re-load the data from the chi-square testing\n",
    "results = OrderedDict()\n",
    "for key, f in zip(samples, founders):\n",
    "    results[key] = pd.read_csv('interchromosomal_linkage_analysis/{}_{}_interchromosomal_ld.tsv'.format(key, f),\n",
    "                               sep='\\t', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM_a</th>\n",
       "      <th>POS_a</th>\n",
       "      <th>CHROM_b</th>\n",
       "      <th>POS_b</th>\n",
       "      <th>AA_obs</th>\n",
       "      <th>AB_obs</th>\n",
       "      <th>BA_obs</th>\n",
       "      <th>BB_obs</th>\n",
       "      <th>chisquare</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr01</td>\n",
       "      <td>6110684</td>\n",
       "      <td>chr02</td>\n",
       "      <td>891366</td>\n",
       "      <td>37</td>\n",
       "      <td>38</td>\n",
       "      <td>36</td>\n",
       "      <td>38</td>\n",
       "      <td>2.894737</td>\n",
       "      <td>0.408141</td>\n",
       "      <td>0.424731</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr01</td>\n",
       "      <td>6110684</td>\n",
       "      <td>chr02</td>\n",
       "      <td>1920981</td>\n",
       "      <td>35</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>35</td>\n",
       "      <td>3.163743</td>\n",
       "      <td>0.367062</td>\n",
       "      <td>0.387602</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr01</td>\n",
       "      <td>6110684</td>\n",
       "      <td>chr02</td>\n",
       "      <td>2404890</td>\n",
       "      <td>31</td>\n",
       "      <td>45</td>\n",
       "      <td>40</td>\n",
       "      <td>33</td>\n",
       "      <td>5.748538</td>\n",
       "      <td>0.124506</td>\n",
       "      <td>0.161255</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr01</td>\n",
       "      <td>6110684</td>\n",
       "      <td>chr02</td>\n",
       "      <td>2672932</td>\n",
       "      <td>32</td>\n",
       "      <td>43</td>\n",
       "      <td>40</td>\n",
       "      <td>33</td>\n",
       "      <td>5.105263</td>\n",
       "      <td>0.164250</td>\n",
       "      <td>0.199813</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr01</td>\n",
       "      <td>6110684</td>\n",
       "      <td>chr02</td>\n",
       "      <td>2687729</td>\n",
       "      <td>32</td>\n",
       "      <td>43</td>\n",
       "      <td>40</td>\n",
       "      <td>33</td>\n",
       "      <td>5.105263</td>\n",
       "      <td>0.164250</td>\n",
       "      <td>0.199813</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr01</td>\n",
       "      <td>6110684</td>\n",
       "      <td>chr02</td>\n",
       "      <td>2741043</td>\n",
       "      <td>31</td>\n",
       "      <td>45</td>\n",
       "      <td>40</td>\n",
       "      <td>33</td>\n",
       "      <td>5.748538</td>\n",
       "      <td>0.124506</td>\n",
       "      <td>0.161255</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr01</td>\n",
       "      <td>6110684</td>\n",
       "      <td>chr02</td>\n",
       "      <td>2827256</td>\n",
       "      <td>30</td>\n",
       "      <td>46</td>\n",
       "      <td>40</td>\n",
       "      <td>33</td>\n",
       "      <td>6.450292</td>\n",
       "      <td>0.091644</td>\n",
       "      <td>0.129735</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr01</td>\n",
       "      <td>6110684</td>\n",
       "      <td>chr02</td>\n",
       "      <td>3286095</td>\n",
       "      <td>33</td>\n",
       "      <td>44</td>\n",
       "      <td>40</td>\n",
       "      <td>33</td>\n",
       "      <td>4.660819</td>\n",
       "      <td>0.198386</td>\n",
       "      <td>0.232054</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr01</td>\n",
       "      <td>6110684</td>\n",
       "      <td>chr02</td>\n",
       "      <td>3349933</td>\n",
       "      <td>33</td>\n",
       "      <td>44</td>\n",
       "      <td>39</td>\n",
       "      <td>34</td>\n",
       "      <td>4.380117</td>\n",
       "      <td>0.223236</td>\n",
       "      <td>0.255650</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr01</td>\n",
       "      <td>6110684</td>\n",
       "      <td>chr02</td>\n",
       "      <td>3463117</td>\n",
       "      <td>33</td>\n",
       "      <td>44</td>\n",
       "      <td>40</td>\n",
       "      <td>32</td>\n",
       "      <td>5.140351</td>\n",
       "      <td>0.161804</td>\n",
       "      <td>0.197259</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM_a    POS_a CHROM_b    POS_b  AA_obs  AB_obs  BA_obs  BB_obs  \\\n",
       "0   chr01  6110684   chr02   891366      37      38      36      38   \n",
       "1   chr01  6110684   chr02  1920981      35      40      40      35   \n",
       "2   chr01  6110684   chr02  2404890      31      45      40      33   \n",
       "3   chr01  6110684   chr02  2672932      32      43      40      33   \n",
       "4   chr01  6110684   chr02  2687729      32      43      40      33   \n",
       "5   chr01  6110684   chr02  2741043      31      45      40      33   \n",
       "6   chr01  6110684   chr02  2827256      30      46      40      33   \n",
       "7   chr01  6110684   chr02  3286095      33      44      40      33   \n",
       "8   chr01  6110684   chr02  3349933      33      44      39      34   \n",
       "9   chr01  6110684   chr02  3463117      33      44      40      32   \n",
       "\n",
       "   chisquare    pvalue    qvalue  significant  \n",
       "0   2.894737  0.408141  0.424731        False  \n",
       "1   3.163743  0.367062  0.387602        False  \n",
       "2   5.748538  0.124506  0.161255        False  \n",
       "3   5.105263  0.164250  0.199813        False  \n",
       "4   5.105263  0.164250  0.199813        False  \n",
       "5   5.748538  0.124506  0.161255        False  \n",
       "6   6.450292  0.091644  0.129735        False  \n",
       "7   4.660819  0.198386  0.232054        False  \n",
       "8   4.380117  0.223236  0.255650        False  \n",
       "9   5.140351  0.161804  0.197259        False  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Did things get loaded?\n",
    "results['N01'][0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Great! Now we must filter for alpha < 0.001 significance\n",
    "significant_results = OrderedDict()\n",
    "for key in samples:\n",
    "    significant_results[key] = results[key].loc[(results[key].significant == True) & (results[key].qvalue < 0.001), :].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM_a</th>\n",
       "      <th>POS_a</th>\n",
       "      <th>CHROM_b</th>\n",
       "      <th>POS_b</th>\n",
       "      <th>AA_obs</th>\n",
       "      <th>AB_obs</th>\n",
       "      <th>BA_obs</th>\n",
       "      <th>BB_obs</th>\n",
       "      <th>chisquare</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18387</th>\n",
       "      <td>chr01</td>\n",
       "      <td>20053392</td>\n",
       "      <td>chr02</td>\n",
       "      <td>31482718</td>\n",
       "      <td>36</td>\n",
       "      <td>24</td>\n",
       "      <td>33</td>\n",
       "      <td>64</td>\n",
       "      <td>22.076023</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>0.000523</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18388</th>\n",
       "      <td>chr01</td>\n",
       "      <td>20053392</td>\n",
       "      <td>chr02</td>\n",
       "      <td>32661204</td>\n",
       "      <td>34</td>\n",
       "      <td>27</td>\n",
       "      <td>32</td>\n",
       "      <td>64</td>\n",
       "      <td>20.859649</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>0.000865</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18389</th>\n",
       "      <td>chr01</td>\n",
       "      <td>20053392</td>\n",
       "      <td>chr02</td>\n",
       "      <td>32676129</td>\n",
       "      <td>34</td>\n",
       "      <td>27</td>\n",
       "      <td>32</td>\n",
       "      <td>64</td>\n",
       "      <td>20.859649</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>0.000865</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18623</th>\n",
       "      <td>chr01</td>\n",
       "      <td>20103165</td>\n",
       "      <td>chr02</td>\n",
       "      <td>31482718</td>\n",
       "      <td>36</td>\n",
       "      <td>24</td>\n",
       "      <td>33</td>\n",
       "      <td>64</td>\n",
       "      <td>22.076023</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>0.000523</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18624</th>\n",
       "      <td>chr01</td>\n",
       "      <td>20103165</td>\n",
       "      <td>chr02</td>\n",
       "      <td>32661204</td>\n",
       "      <td>34</td>\n",
       "      <td>27</td>\n",
       "      <td>32</td>\n",
       "      <td>64</td>\n",
       "      <td>20.859649</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>0.000865</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18625</th>\n",
       "      <td>chr01</td>\n",
       "      <td>20103165</td>\n",
       "      <td>chr02</td>\n",
       "      <td>32676129</td>\n",
       "      <td>34</td>\n",
       "      <td>27</td>\n",
       "      <td>32</td>\n",
       "      <td>64</td>\n",
       "      <td>20.859649</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>0.000865</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18859</th>\n",
       "      <td>chr01</td>\n",
       "      <td>20144421</td>\n",
       "      <td>chr02</td>\n",
       "      <td>31482718</td>\n",
       "      <td>36</td>\n",
       "      <td>24</td>\n",
       "      <td>33</td>\n",
       "      <td>64</td>\n",
       "      <td>22.076023</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>0.000523</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18860</th>\n",
       "      <td>chr01</td>\n",
       "      <td>20144421</td>\n",
       "      <td>chr02</td>\n",
       "      <td>32661204</td>\n",
       "      <td>34</td>\n",
       "      <td>27</td>\n",
       "      <td>32</td>\n",
       "      <td>64</td>\n",
       "      <td>20.859649</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>0.000865</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18861</th>\n",
       "      <td>chr01</td>\n",
       "      <td>20144421</td>\n",
       "      <td>chr02</td>\n",
       "      <td>32676129</td>\n",
       "      <td>34</td>\n",
       "      <td>27</td>\n",
       "      <td>32</td>\n",
       "      <td>64</td>\n",
       "      <td>20.859649</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>0.000865</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19095</th>\n",
       "      <td>chr01</td>\n",
       "      <td>20154857</td>\n",
       "      <td>chr02</td>\n",
       "      <td>31482718</td>\n",
       "      <td>36</td>\n",
       "      <td>24</td>\n",
       "      <td>33</td>\n",
       "      <td>64</td>\n",
       "      <td>22.076023</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>0.000523</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM_a     POS_a CHROM_b     POS_b  AA_obs  AB_obs  BA_obs  BB_obs  \\\n",
       "18387   chr01  20053392   chr02  31482718      36      24      33      64   \n",
       "18388   chr01  20053392   chr02  32661204      34      27      32      64   \n",
       "18389   chr01  20053392   chr02  32676129      34      27      32      64   \n",
       "18623   chr01  20103165   chr02  31482718      36      24      33      64   \n",
       "18624   chr01  20103165   chr02  32661204      34      27      32      64   \n",
       "18625   chr01  20103165   chr02  32676129      34      27      32      64   \n",
       "18859   chr01  20144421   chr02  31482718      36      24      33      64   \n",
       "18860   chr01  20144421   chr02  32661204      34      27      32      64   \n",
       "18861   chr01  20144421   chr02  32676129      34      27      32      64   \n",
       "19095   chr01  20154857   chr02  31482718      36      24      33      64   \n",
       "\n",
       "       chisquare    pvalue    qvalue  significant  \n",
       "18387  22.076023  0.000063  0.000523         True  \n",
       "18388  20.859649  0.000113  0.000865         True  \n",
       "18389  20.859649  0.000113  0.000865         True  \n",
       "18623  22.076023  0.000063  0.000523         True  \n",
       "18624  20.859649  0.000113  0.000865         True  \n",
       "18625  20.859649  0.000113  0.000865         True  \n",
       "18859  22.076023  0.000063  0.000523         True  \n",
       "18860  20.859649  0.000113  0.000865         True  \n",
       "18861  20.859649  0.000113  0.000865         True  \n",
       "19095  22.076023  0.000063  0.000523         True  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "significant_results['N01'][0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now, we can record which genotype is being skewed for by examining the max of the four columns for each row\n",
    "for key in samples:\n",
    "    significant_results[key]['skew'] = significant_results[key].iloc[:, 4:8].idxmax(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM_a</th>\n",
       "      <th>POS_a</th>\n",
       "      <th>CHROM_b</th>\n",
       "      <th>POS_b</th>\n",
       "      <th>AA_obs</th>\n",
       "      <th>AB_obs</th>\n",
       "      <th>BA_obs</th>\n",
       "      <th>BB_obs</th>\n",
       "      <th>chisquare</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>significant</th>\n",
       "      <th>skew</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18387</th>\n",
       "      <td>chr01</td>\n",
       "      <td>20053392</td>\n",
       "      <td>chr02</td>\n",
       "      <td>31482718</td>\n",
       "      <td>36</td>\n",
       "      <td>24</td>\n",
       "      <td>33</td>\n",
       "      <td>64</td>\n",
       "      <td>22.076023</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>0.000523</td>\n",
       "      <td>True</td>\n",
       "      <td>BB_obs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18388</th>\n",
       "      <td>chr01</td>\n",
       "      <td>20053392</td>\n",
       "      <td>chr02</td>\n",
       "      <td>32661204</td>\n",
       "      <td>34</td>\n",
       "      <td>27</td>\n",
       "      <td>32</td>\n",
       "      <td>64</td>\n",
       "      <td>20.859649</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>0.000865</td>\n",
       "      <td>True</td>\n",
       "      <td>BB_obs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18389</th>\n",
       "      <td>chr01</td>\n",
       "      <td>20053392</td>\n",
       "      <td>chr02</td>\n",
       "      <td>32676129</td>\n",
       "      <td>34</td>\n",
       "      <td>27</td>\n",
       "      <td>32</td>\n",
       "      <td>64</td>\n",
       "      <td>20.859649</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>0.000865</td>\n",
       "      <td>True</td>\n",
       "      <td>BB_obs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18623</th>\n",
       "      <td>chr01</td>\n",
       "      <td>20103165</td>\n",
       "      <td>chr02</td>\n",
       "      <td>31482718</td>\n",
       "      <td>36</td>\n",
       "      <td>24</td>\n",
       "      <td>33</td>\n",
       "      <td>64</td>\n",
       "      <td>22.076023</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>0.000523</td>\n",
       "      <td>True</td>\n",
       "      <td>BB_obs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18624</th>\n",
       "      <td>chr01</td>\n",
       "      <td>20103165</td>\n",
       "      <td>chr02</td>\n",
       "      <td>32661204</td>\n",
       "      <td>34</td>\n",
       "      <td>27</td>\n",
       "      <td>32</td>\n",
       "      <td>64</td>\n",
       "      <td>20.859649</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>0.000865</td>\n",
       "      <td>True</td>\n",
       "      <td>BB_obs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18625</th>\n",
       "      <td>chr01</td>\n",
       "      <td>20103165</td>\n",
       "      <td>chr02</td>\n",
       "      <td>32676129</td>\n",
       "      <td>34</td>\n",
       "      <td>27</td>\n",
       "      <td>32</td>\n",
       "      <td>64</td>\n",
       "      <td>20.859649</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>0.000865</td>\n",
       "      <td>True</td>\n",
       "      <td>BB_obs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18859</th>\n",
       "      <td>chr01</td>\n",
       "      <td>20144421</td>\n",
       "      <td>chr02</td>\n",
       "      <td>31482718</td>\n",
       "      <td>36</td>\n",
       "      <td>24</td>\n",
       "      <td>33</td>\n",
       "      <td>64</td>\n",
       "      <td>22.076023</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>0.000523</td>\n",
       "      <td>True</td>\n",
       "      <td>BB_obs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18860</th>\n",
       "      <td>chr01</td>\n",
       "      <td>20144421</td>\n",
       "      <td>chr02</td>\n",
       "      <td>32661204</td>\n",
       "      <td>34</td>\n",
       "      <td>27</td>\n",
       "      <td>32</td>\n",
       "      <td>64</td>\n",
       "      <td>20.859649</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>0.000865</td>\n",
       "      <td>True</td>\n",
       "      <td>BB_obs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18861</th>\n",
       "      <td>chr01</td>\n",
       "      <td>20144421</td>\n",
       "      <td>chr02</td>\n",
       "      <td>32676129</td>\n",
       "      <td>34</td>\n",
       "      <td>27</td>\n",
       "      <td>32</td>\n",
       "      <td>64</td>\n",
       "      <td>20.859649</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>0.000865</td>\n",
       "      <td>True</td>\n",
       "      <td>BB_obs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19095</th>\n",
       "      <td>chr01</td>\n",
       "      <td>20154857</td>\n",
       "      <td>chr02</td>\n",
       "      <td>31482718</td>\n",
       "      <td>36</td>\n",
       "      <td>24</td>\n",
       "      <td>33</td>\n",
       "      <td>64</td>\n",
       "      <td>22.076023</td>\n",
       "      <td>0.000063</td>\n",
       "      <td>0.000523</td>\n",
       "      <td>True</td>\n",
       "      <td>BB_obs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM_a     POS_a CHROM_b     POS_b  AA_obs  AB_obs  BA_obs  BB_obs  \\\n",
       "18387   chr01  20053392   chr02  31482718      36      24      33      64   \n",
       "18388   chr01  20053392   chr02  32661204      34      27      32      64   \n",
       "18389   chr01  20053392   chr02  32676129      34      27      32      64   \n",
       "18623   chr01  20103165   chr02  31482718      36      24      33      64   \n",
       "18624   chr01  20103165   chr02  32661204      34      27      32      64   \n",
       "18625   chr01  20103165   chr02  32676129      34      27      32      64   \n",
       "18859   chr01  20144421   chr02  31482718      36      24      33      64   \n",
       "18860   chr01  20144421   chr02  32661204      34      27      32      64   \n",
       "18861   chr01  20144421   chr02  32676129      34      27      32      64   \n",
       "19095   chr01  20154857   chr02  31482718      36      24      33      64   \n",
       "\n",
       "       chisquare    pvalue    qvalue  significant    skew  \n",
       "18387  22.076023  0.000063  0.000523         True  BB_obs  \n",
       "18388  20.859649  0.000113  0.000865         True  BB_obs  \n",
       "18389  20.859649  0.000113  0.000865         True  BB_obs  \n",
       "18623  22.076023  0.000063  0.000523         True  BB_obs  \n",
       "18624  20.859649  0.000113  0.000865         True  BB_obs  \n",
       "18625  20.859649  0.000113  0.000865         True  BB_obs  \n",
       "18859  22.076023  0.000063  0.000523         True  BB_obs  \n",
       "18860  20.859649  0.000113  0.000865         True  BB_obs  \n",
       "18861  20.859649  0.000113  0.000865         True  BB_obs  \n",
       "19095  22.076023  0.000063  0.000523         True  BB_obs  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "significant_results['N01'][0:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Next, we have to re-write the Interchromosomal LD binning to also account for coupling AA, coupling BB, or repulsion\n",
    "Since coupling can be skewed either `AA` or `BB` and heterozygous skew is not directional (`AB` or `BA`), we have 3 possible colorings for the links. So when extending bins, we must not only account for distance, but also type of skew."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We want to obtain coords for all loci showing interchromosomal LD and account for coupling vs repulasion\n",
    "max_distance = 5000000\n",
    "loci = OrderedDict()\n",
    "for key in samples:\n",
    "    rows = significant_results[key].itertuples(index=False)\n",
    "    # Prime zee pump!\n",
    "    try:\n",
    "        prev_row = next(rows)\n",
    "    except StopIteration:\n",
    "        # Iterator was empty so we call continue to move to the next sample\n",
    "        continue\n",
    "    # OK if we got here this sample should get a key in the OrderedDict()\n",
    "    loci[key] = list()\n",
    "    # Set the initial locus coords\n",
    "    block = {'CHROM_a': prev_row.CHROM_a, 'POS_a_start': prev_row.POS_a, 'POS_a_end': prev_row.POS_a,\n",
    "             'CHROM_b': prev_row.CHROM_b, 'POS_b_start': prev_row.POS_b, 'POS_b_end': prev_row.POS_b,\n",
    "             'skew': prev_row.skew,}\n",
    "    for curr_row in rows:\n",
    "        if (curr_row.CHROM_a == prev_row.CHROM_a) and (curr_row.CHROM_b == prev_row.CHROM_b):\n",
    "            if (curr_row.POS_a - prev_row.POS_a <= max_distance) and (np.abs(curr_row.POS_b - prev_row.POS_b) <= max_distance) and (curr_row.skew == prev_row.skew):\n",
    "                # Expand your locus to include this marker\n",
    "                # Note: Due to the way the data is sorted the \"a\" is monotonically increasing\n",
    "                block['POS_a_end'] = curr_row.POS_a\n",
    "                # Note: But the \"b\" data increases and then \"resets\" at each new \"a\"\n",
    "                #       This is due to the nature of how the combinatoric comparisons were performed.\n",
    "                #       Thus we need to ensure that we should actually be setting new start or end coordinates\n",
    "                #       for the \"b\" side of this block.\n",
    "                if curr_row.POS_b > block['POS_b_end']:\n",
    "                    block['POS_b_end'] = curr_row.POS_b\n",
    "                if curr_row.POS_b < block['POS_b_start']:\n",
    "                    block['POS_b_start'] = curr_row.POS_b\n",
    "            else:\n",
    "                # We just done entered a new locus dudebro!\n",
    "                # Add the coord-pair for the locus to the list of coord-pairs for the sample\n",
    "                #loci[key].append([start_pos.CHROM_a, np.mean([start_pos.POS_a, end_pos.POS_a]),\n",
    "                #                       start_pos.CHROM_b, np.mean([start_pos.POS_b, end_pos.POS_b])])\n",
    "                loci[key].append([block['CHROM_a'], block['POS_a_start'], block['POS_a_end'],\n",
    "                                  block['CHROM_b'], block['POS_b_start'], block['POS_b_end'],\n",
    "                                  block['skew']])\n",
    "                # Now reset the start_pos & end_pos, because we are at a new locus\n",
    "                block['CHROM_a'] = curr_row.CHROM_a\n",
    "                block['POS_a_start'] = curr_row.POS_a\n",
    "                block['POS_a_end'] = curr_row.POS_a\n",
    "                block['CHROM_b'] = curr_row.CHROM_b\n",
    "                block['POS_b_start'] = curr_row.POS_b\n",
    "                block['POS_b_end'] = curr_row.POS_b\n",
    "                block['skew'] = curr_row.skew\n",
    "        else:\n",
    "            # Walked into a new chromosome, add the coord-pair for the locus to the list of coord-pairs for the sample\n",
    "            #loci[key].append([start_pos.CHROM_a, np.mean([start_pos.POS_a, end_pos.POS_a]),\n",
    "            #                       start_pos.CHROM_b, np.mean([start_pos.POS_b, end_pos.POS_b])])\n",
    "            loci[key].append([block['CHROM_a'], block['POS_a_start'], block['POS_a_end'],\n",
    "                              block['CHROM_b'], block['POS_b_start'], block['POS_b_end'],\n",
    "                              block['skew']])\n",
    "            # Now reset the start & end_pos, because we are at a new locus\n",
    "            block['CHROM_a'] = curr_row.CHROM_a\n",
    "            block['POS_a_start'] = curr_row.POS_a\n",
    "            block['POS_a_end'] = curr_row.POS_a\n",
    "            block['CHROM_b'] = curr_row.CHROM_b\n",
    "            block['POS_b_start'] = curr_row.POS_b\n",
    "            block['POS_b_end'] = curr_row.POS_b\n",
    "            block['skew'] = curr_row.skew\n",
    "        # No matter what else happened up above, the curr_row is now the prev_row\n",
    "        prev_row = curr_row\n",
    "    # Now we're done with this sample, so let's make this list of lists into a DataFrame\n",
    "    loci[key] = pd.DataFrame(data=loci[key], columns=['CHROM_a', 'POS_a_start', 'POS_a_end',\n",
    "                                                                 'CHROM_b', 'POS_b_start', 'POS_b_end', 'skew'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM_a</th>\n",
       "      <th>POS_a_start</th>\n",
       "      <th>POS_a_end</th>\n",
       "      <th>CHROM_b</th>\n",
       "      <th>POS_b_start</th>\n",
       "      <th>POS_b_end</th>\n",
       "      <th>skew</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr01</td>\n",
       "      <td>20053392</td>\n",
       "      <td>22113938</td>\n",
       "      <td>chr02</td>\n",
       "      <td>31482718</td>\n",
       "      <td>34030257</td>\n",
       "      <td>BB_obs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr01</td>\n",
       "      <td>22148052</td>\n",
       "      <td>22148052</td>\n",
       "      <td>chr02</td>\n",
       "      <td>18646610</td>\n",
       "      <td>18686916</td>\n",
       "      <td>BB_obs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr01</td>\n",
       "      <td>22148052</td>\n",
       "      <td>22148052</td>\n",
       "      <td>chr02</td>\n",
       "      <td>29563080</td>\n",
       "      <td>34030257</td>\n",
       "      <td>BB_obs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr01</td>\n",
       "      <td>22148902</td>\n",
       "      <td>22148902</td>\n",
       "      <td>chr02</td>\n",
       "      <td>18646610</td>\n",
       "      <td>18686916</td>\n",
       "      <td>BB_obs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr01</td>\n",
       "      <td>22148902</td>\n",
       "      <td>22277753</td>\n",
       "      <td>chr02</td>\n",
       "      <td>29563080</td>\n",
       "      <td>34030257</td>\n",
       "      <td>BB_obs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr01</td>\n",
       "      <td>22313907</td>\n",
       "      <td>22313907</td>\n",
       "      <td>chr02</td>\n",
       "      <td>18646610</td>\n",
       "      <td>18686916</td>\n",
       "      <td>BB_obs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr01</td>\n",
       "      <td>22313907</td>\n",
       "      <td>22313907</td>\n",
       "      <td>chr02</td>\n",
       "      <td>29370785</td>\n",
       "      <td>34030257</td>\n",
       "      <td>BB_obs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr01</td>\n",
       "      <td>22319807</td>\n",
       "      <td>22319807</td>\n",
       "      <td>chr02</td>\n",
       "      <td>18646610</td>\n",
       "      <td>18686916</td>\n",
       "      <td>BB_obs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr01</td>\n",
       "      <td>22319807</td>\n",
       "      <td>22336584</td>\n",
       "      <td>chr02</td>\n",
       "      <td>29370785</td>\n",
       "      <td>34030257</td>\n",
       "      <td>BB_obs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr01</td>\n",
       "      <td>22352813</td>\n",
       "      <td>22352813</td>\n",
       "      <td>chr02</td>\n",
       "      <td>18646610</td>\n",
       "      <td>18686916</td>\n",
       "      <td>BB_obs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM_a  POS_a_start  POS_a_end CHROM_b  POS_b_start  POS_b_end    skew\n",
       "0   chr01     20053392   22113938   chr02     31482718   34030257  BB_obs\n",
       "1   chr01     22148052   22148052   chr02     18646610   18686916  BB_obs\n",
       "2   chr01     22148052   22148052   chr02     29563080   34030257  BB_obs\n",
       "3   chr01     22148902   22148902   chr02     18646610   18686916  BB_obs\n",
       "4   chr01     22148902   22277753   chr02     29563080   34030257  BB_obs\n",
       "5   chr01     22313907   22313907   chr02     18646610   18686916  BB_obs\n",
       "6   chr01     22313907   22313907   chr02     29370785   34030257  BB_obs\n",
       "7   chr01     22319807   22319807   chr02     18646610   18686916  BB_obs\n",
       "8   chr01     22319807   22336584   chr02     29370785   34030257  BB_obs\n",
       "9   chr01     22352813   22352813   chr02     18646610   18686916  BB_obs"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci['N01'][0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(15989, 7)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci['N01'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in loci.keys():\n",
    "    loci[key].to_csv('interchromosomal_linkage_analysis/{}_ICLD_gtskew_5M_windows.tsv'.format(key), sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Searching for lethal combinations\n",
    "To identify lethal gene combinations, we can examine instances where homozygosity for either parent's markers are not seen within a population. This is _prima facie_ evidence that such a combination is non-viable (i.e. - lethal) because we should expect to see pairs of non-linked homozygous loci appear in even proportion to Mendelian segregation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "loci_of_interest = OrderedDict()\n",
    "for key in samples:\n",
    "    loci_of_interest[key] = significant_results[key].loc[(significant_results[key].AA_obs == 0) | (significant_results[key].BB_obs == 0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's remove keys from the dictionary where the shape of the dataframe is 0 rows, \n",
    "# meaning there were no instances of potentially lethal genotypes seen between loci.\n",
    "for key in samples:\n",
    "    if loci_of_interest[key].shape[0] == 0:\n",
    "        loci_of_interest.pop(key, None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "odict_keys(['N01', 'N07', 'N08', 'N09', 'N14', 'N16', 'N17'])"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# See which samples remain from the original set\n",
    "loci_of_interest.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N01 skew values present {'AA_obs', 'AB_obs'}.\n",
      "N07 skew values present {'AA_obs', 'AB_obs'}.\n",
      "N08 skew values present {'AA_obs', 'AB_obs'}.\n",
      "N09 skew values present {'AA_obs', 'AB_obs'}.\n",
      "N14 skew values present {'AA_obs', 'AB_obs'}.\n",
      "N16 skew values present {'AA_obs', 'AB_obs'}.\n",
      "N17 skew values present {'AA_obs', 'AB_obs'}.\n"
     ]
    }
   ],
   "source": [
    "# Tell me which direction the genome skews for each remaining sample\n",
    "for key in loci_of_interest:\n",
    "    print('{} skew values present {}.'.format(key, set(loci_of_interest['N01']['skew'])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
