{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Author: Dan Shea  \n",
    "Date: 2019.08.08  \n",
    "\n",
    "# VCFstats\n",
    "`VCFstats` will load a VCF file as a `pandas` dataframe and calculate statistical metrics that may be of use when examining variant data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "import numpy as np\n",
    "import scipy.stats as stats\n",
    "import os.path\n",
    "import os\n",
    "import re\n",
    "from collections import OrderedDict\n",
    "import gzip\n",
    "import statsmodels.stats.multitest as multitest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Error(Exception):\n",
    "    \"\"\"Base class for exceptions in this module.\n",
    "    Attributes:\n",
    "    expression -- input expression in which the error occurred\n",
    "    message -- explanation of the error\n",
    "    \"\"\"\n",
    "    def __init__(self, expression, message):\n",
    "        self.expression = expression\n",
    "        self.message = message\n",
    "\n",
    "        \n",
    "class PloidyError(Error):\n",
    "    \"\"\"Exception raised for ploidy errors in the input.\"\"\"\n",
    "    pass\n",
    "\n",
    "\n",
    "class HeterozygousParentError(Error):\n",
    "    pass\n",
    "\n",
    "\n",
    "class NoAlleleFrequenciesError(Error):\n",
    "    pass\n",
    "\n",
    "        \n",
    "class VCF:\n",
    "    def __is_double_hash__(self, line):\n",
    "        match = re.match('^##', line)\n",
    "        if match:\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "        \n",
    "    def __is_single_hash__(self, line):\n",
    "        match = re.match('^#[^#]', line)\n",
    "        if match:\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "    \n",
    "    def __is_key_value_pair__(self, line):\n",
    "        match = re.match('^(.*?)=(.*)$', line)\n",
    "        if match:\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "    \n",
    "    def __isref__(self, val):\n",
    "        match = re.match('0[|/]0', val)\n",
    "        if match:\n",
    "            return True\n",
    "        else:\n",
    "            match = re.match('1[|/]1', val)\n",
    "            if match:\n",
    "                return False\n",
    "            else:\n",
    "                match = re.match('0[|/]1', val)\n",
    "                if match:\n",
    "                    raise HeterozygousParentError(val, 'Parent contains a heterozygous call at this loci.')\n",
    "                else:\n",
    "                    match = re.match('1[|/]0', val)\n",
    "                    if match:\n",
    "                        raise HeterozygousParentError(val, 'Parent contains a heterozygous call at this loci.')\n",
    "                    else:\n",
    "                        raise PloidyError(val, 'Genotype does not appear to be diploid.')\n",
    "    \n",
    "    def __init__(self, vcf_filename, reference_filename=None):\n",
    "        # Internal VCF data\n",
    "        self.__colnames__ = list()\n",
    "        self.__samples__ = list()\n",
    "        self.__hasFORMAT__ = False\n",
    "        self.__expected_frequencies__ = np.array([0.5, 0.5]) # Expected Mendelian frequencies\n",
    "        # Attributes\n",
    "        self.meta = OrderedDict()       # meta-data in the VCF (key-value pairs)\n",
    "        self.data = pd.DataFrame()      # Pandas dataframe of the genotyping data\n",
    "        self.num_records = 0            # How many records were read in at the time of intialization\n",
    "        self.num_samples = 0            # How many samples are in the file? (Not counting the parent samples)\n",
    "        self.vcf_filename = None        # The VCF filename\n",
    "        self.reference_filename = None  # The reference genome associated with the VCF\n",
    "        self.seqio = None               # If a reference was given, a SeqIO of the reference\n",
    "        self.allele_frequencies = None  # Allele frequencies calculated by calling calculate_allele_frequencies() method\n",
    "        self.dump_prefix = None         # Default filename prefix for the dump() method\n",
    "        self.chisquare_pvalues = None   # p-values of a chi-square goodness of fit test against calculated allele frequencies\n",
    "        self.chisquare_qvalues = None   # Bejamini-Hochberg FDR corrected p-values of a chi-square goodness of fit test against\n",
    "                                        # calculated allele frequencies.\n",
    "        \n",
    "        # Parse the VCF\n",
    "        if os.path.isfile(vcf_filename):\n",
    "            self.vcf_filename = vcf_filename\n",
    "            with gzip.open(vcf_filename, 'rt') as vcfh:\n",
    "                header_was_read = False\n",
    "                vcf_data = list()\n",
    "                for line in vcfh:\n",
    "                    line = line.strip()\n",
    "                    # Double-hash lines are meta lines that consist of key-value pairs\n",
    "                    # i.e. - ##key=value\n",
    "                    # If we didn't already see the header check if this line is a key-value pair\n",
    "                    if not header_was_read and self.__is_double_hash__(line):\n",
    "                        line = line.lstrip('#')\n",
    "                        if self.__is_key_value_pair__(line):\n",
    "                            match = re.match('^(.*?)=(.*)$', line)\n",
    "                            self.meta[match.group(0)] = match.group(1)\n",
    "                    # The single hash line is the header line defining the mandatory eight columns:\n",
    "                    #  1. CHROM  - chromosome: An identifier from the reference genome\n",
    "                    #  2. POS    - position: The reference position, with the 1st base having position 1.\n",
    "                    #  3. ID     - identifier: Semi-colon separated list of unique identifiers where available.\n",
    "                    #  4. REF    - reference base(s): Each base must be one of A,C,G,T,N (case insensitive).\n",
    "                    #  5. ALT    - alternate base(s): Comma separated list of alternate non-reference alleles. (A,C,G,T,N,*)\n",
    "                    #  6. QUAL   - quality: Phred-scaled quality score for the assertion made in ALT.\n",
    "                    #  7. FILTER - filter status: PASS if this position has passed all filters, i.e. a call is made at\n",
    "                    #             this position.\n",
    "                    #  8. INFO   - additional information: (String, no white-space, semi-colons, or equals-signs permitted;\n",
    "                    #             commas are permitted only as delimiters for lists of values)\n",
    "                    #  9. FORMAT - If genotype information is present, then the same types of data must be present for all\n",
    "                    #             samples. First a FORMAT field is given specifying the data types and order (colon-separated\n",
    "                    #             alphanumeric String).\n",
    "                    # 10. SAMPLE - ALL columns after these columns are unique sample identifiers.\n",
    "                    \n",
    "                    # ***** THE HEADER LINE FIELDS ARE TAB-DELIMITED *****\n",
    "                    \n",
    "                    # If we didn't already see the header, check if this line is the header\n",
    "                    elif not header_was_read and self.__is_single_hash__(line):\n",
    "                        header_was_read = True\n",
    "                        line = line.lstrip('#')\n",
    "                        fields = line.split('\\t')\n",
    "                        if fields[8] == 'FORMAT':\n",
    "                            self.__hasFORMAT__ = True\n",
    "                            self.__colnames__ = fields[0:9]\n",
    "                            self.__samples__ = fields[9:]\n",
    "                        else:\n",
    "                            self.__colnames__ = fields[0:8]\n",
    "                            self.__samples__ = fields[8:]\n",
    "                        \n",
    "                        # We subtract 2 here because HITOMEBORE and the FOUNDER are also in the file\n",
    "                        self.num_samples = len(self.__samples__) - 2 \n",
    "                    \n",
    "                    # If we already saw the header, start reading in the data\n",
    "                    elif header_was_read:\n",
    "                        fields = line.split('\\t')\n",
    "                        # Convert non-string data to the appropriate type\n",
    "                        fields[1] = int(fields[1])\n",
    "                        # Because variant calls are performed with respect to a reference sequence\n",
    "                        # We must account for whichi parent matches the reference and which differs.\n",
    "                        # Homozygous Hitomebore (dependent upon whether it matches the REF or the ALT)\n",
    "                        # Homozygous Founder (dependent upon whether it matches the REF or the ALT)\n",
    "                        # Heterozygous (0|1) or (1|0)\n",
    "                        start = None\n",
    "                        if self.__hasFORMAT__:\n",
    "                            start = 9\n",
    "                        else:\n",
    "                            start = 8\n",
    "                        # Analyze parents\n",
    "                        hitomebore = fields[start]\n",
    "                        fields[start] = 'A'\n",
    "                        founder = fields[start+1]\n",
    "                        fields[start+1] = 'B'\n",
    "                        \n",
    "                        # But who was phone^H^H^H^H^H ref?\n",
    "                        if self.__isref__(hitomebore):\n",
    "                            aref = True\n",
    "                        else:\n",
    "                            aref = False\n",
    "                        \n",
    "                        # A little rendundant, but I want to ensure there's no weirdness in the founder genotypes either\n",
    "                        if self.__isref__(founder):\n",
    "                            aref = False\n",
    "                        else:\n",
    "                            aref = True\n",
    "                        \n",
    "                        for i in range(start+2, len(fields)):\n",
    "                            val = fields[i]\n",
    "                            match = re.match('0[|/]0', val)\n",
    "                            if match:\n",
    "                                fields[i] = 'A' if aref else 'B'\n",
    "                            else:\n",
    "                                match = re.match('1[|/]1', val)\n",
    "                                if match:\n",
    "                                    fields[i] = 'B' if aref else 'A'\n",
    "                                else:\n",
    "                                    match = re.match('0[|/]1', val)\n",
    "                                    if match:\n",
    "                                        fields[i] = 'H'\n",
    "                                    else:\n",
    "                                        match = re.match('1[|/]0', val)\n",
    "                                        if match:\n",
    "                                            fields[i] = 'H'\n",
    "                                        else:\n",
    "                                            raise PloidyError(val, 'Genotype does not appear to be diploid.')\n",
    "                                            \n",
    "                        vcf_data.append(fields)\n",
    "                self.data = pd.DataFrame(data=vcf_data, columns=self.__colnames__+self.__samples__)\n",
    "                self.num_records = self.data.shape[0]\n",
    "        \n",
    "        # Parse the reference genome (if provided)\n",
    "        if reference_filename:\n",
    "            if os.path.isfile(reference_filename):\n",
    "                seqio = SeqIO.parse(reference_filename)\n",
    "                self.reference_filename = reference_filename\n",
    "                self.seqio = seqio\n",
    "        \n",
    "        # Define a filename for the dump() method\n",
    "        _tmp = os.path.basename(self.vcf_filename)\n",
    "        _tmp = _tmp.split(os.path.extsep)\n",
    "        _tmp.pop() # remove the gz extension\n",
    "        _tmp.pop() # remove the vcf extension\n",
    "        _tmp = os.path.extsep.join(_tmp) # re-join\n",
    "        self.dump_prefix = os.path.join(os.path.dirname(self.vcf_filename), _tmp)\n",
    "        del(_tmp)\n",
    "                     \n",
    "        # Note: In Python, class constructor __init__(self, ...) does not return anything\n",
    "\n",
    "    def __calculate_allele_frequency__(self, x):\n",
    "        obs = np.array([0,0])\n",
    "        if self.__hasFORMAT__:\n",
    "            start=11\n",
    "        else:\n",
    "            start=10\n",
    "            \n",
    "        num_obs = len(range(start, len(x)))\n",
    "        \n",
    "        for i in range(start, len(x)):\n",
    "            if x[i] == 'A':\n",
    "                obs[0] += 1\n",
    "            elif x[i] == 'B':\n",
    "                obs[1] += 1\n",
    "        return obs\n",
    "    \n",
    "    def __run_chisquare__(self, x):\n",
    "        return stats.chisquare(f_obs=x, f_exp=sum(x)*self.__expected_frequencies__)\n",
    "    \n",
    "    ###### Public methods ######\n",
    "    \n",
    "    def reset_seqio(self):\n",
    "        '''reset_seqio() resets the seqio generator back to the beginning of the file\n",
    "        The seqio attribute is reset, but no value is returned. Access the seqio via its attribute 'seqio'.\n",
    "        '''\n",
    "        if os.path.isfile(self.reference_filename):\n",
    "            seqio = SeqIO.parse(reference_filename)\n",
    "            self.seqio = seqio\n",
    "            \n",
    "    def calculate_allele_frequencies(self):\n",
    "        '''calculate_allele_frequencies() calculates the frequency of Homozygous parental alleles.\n",
    "        Heterozygous are not used in chi-square testing but may be inferred from the data.\n",
    "        Returns the DataFrame which is also accessible as the attribute 'allele_frequencies'.\n",
    "        '''\n",
    "        self.allele_frequencies = self.data.apply(self.__calculate_allele_frequency__, axis=1)\n",
    "        self.allele_frequencies = pd.DataFrame(data=self.allele_frequencies.values.tolist(),\n",
    "                                               columns=['Homozygous A', 'Homozygous B'])\n",
    "        return self.allele_frequencies\n",
    "    \n",
    "    def chisquare(self):\n",
    "        '''Perform a chi-square goodness of fit on the observed allele frequencies.\n",
    "        Return the p-values and save them in the chisquare_pvalues attribute.\n",
    "        '''\n",
    "        if self.allele_frequencies is not None:\n",
    "            self.chisquare_pvalues = pd.DataFrame(self.allele_frequencies.apply(self.__run_chisquare__, axis=1).values.tolist())\n",
    "            self.allele_frequencies = pd.concat([self.allele_frequencies, self.chisquare_pvalues], axis=1)\n",
    "            # Now we perform FDR correction\n",
    "            sig, qvalue = multitest.fdrcorrection(self.chisquare_pvalues.loc[:, 'pvalue'])\n",
    "            self.chisquare_qvalues = pd.DataFrame(zip(qvalue, sig), columns=['qvalue', 'significant'])\n",
    "            self.allele_frequencies = pd.concat([self.allele_frequencies, self.chisquare_qvalues], axis=1)\n",
    "            return self.allele_frequencies\n",
    "        else:\n",
    "            raise NoAlleleFrequenciesError(None, 'You must run calculate_allele_frequencies() to perform a chi-square test.')\n",
    "    \n",
    "    def dump(self, prefix=None):\n",
    "        '''dump() creates a tsv consisting of the:\n",
    "        Genotype (if present)\n",
    "        Allele frequencies (if present)\n",
    "        Default output is written to the current vcf path and filename (sans .vcf) with an xlsx extension.\n",
    "        May be changed by passing a filename argument.\n",
    "        '''\n",
    "        if not prefix:\n",
    "            prefix = self.dump_prefix\n",
    "        if self.data is not None:\n",
    "            self.data.to_csv(prefix+'_genotypes.tsv', sep='\\t')\n",
    "        if self.allele_frequencies is not None:\n",
    "            self.allele_frequencies.to_csv(prefix+'_allele_frequencies.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = VCF('beagle_output/TEST_DATA/TEST.vcf.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<__main__.VCF at 0x7f5d9a4dc6a0>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>N00_HITOMEBORE</th>\n",
       "      <th>...</th>\n",
       "      <th>N01_164</th>\n",
       "      <th>N01_165</th>\n",
       "      <th>N01_166</th>\n",
       "      <th>N01_167</th>\n",
       "      <th>N01_168</th>\n",
       "      <th>N01_169</th>\n",
       "      <th>N01_170</th>\n",
       "      <th>N01_171</th>\n",
       "      <th>N01_172</th>\n",
       "      <th>N01_173</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr01</td>\n",
       "      <td>1411</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>H</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>H</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr01</td>\n",
       "      <td>6918</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>H</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>H</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr02</td>\n",
       "      <td>9619</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>B</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr02</td>\n",
       "      <td>36401</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>B</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>B</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr03</td>\n",
       "      <td>2137</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>H</td>\n",
       "      <td>B</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr03</td>\n",
       "      <td>198008</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>GT</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>H</td>\n",
       "      <td>B</td>\n",
       "      <td>A</td>\n",
       "      <td>H</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 182 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM     POS ID REF ALT QUAL FILTER INFO FORMAT N00_HITOMEBORE  ...  \\\n",
       "0  chr01    1411  .   A   G    .   PASS    .     GT              A  ...   \n",
       "1  chr01    6918  .   G   T    .   PASS    .     GT              A  ...   \n",
       "2  chr02    9619  .   T   C    .   PASS    .     GT              A  ...   \n",
       "3  chr02   36401  .   G   A    .   PASS    .     GT              A  ...   \n",
       "4  chr03    2137  .   A   G    .   PASS    .     GT              A  ...   \n",
       "5  chr03  198008  .   G   A    .   PASS    .     GT              A  ...   \n",
       "\n",
       "  N01_164 N01_165 N01_166 N01_167 N01_168 N01_169 N01_170 N01_171 N01_172  \\\n",
       "0       A       H       A       A       B       B       H       A       A   \n",
       "1       A       H       A       A       B       B       H       A       A   \n",
       "2       B       A       B       A       B       B       A       A       B   \n",
       "3       B       A       B       A       B       B       A       A       B   \n",
       "4       A       H       B       A       A       A       A       A       A   \n",
       "5       A       H       B       A       H       A       A       A       A   \n",
       "\n",
       "  N01_173  \n",
       "0       A  \n",
       "1       A  \n",
       "2       B  \n",
       "3       B  \n",
       "4       A  \n",
       "5       A  \n",
       "\n",
       "[6 rows x 182 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Homozygous A</th>\n",
       "      <th>Homozygous B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>62</td>\n",
       "      <td>96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>62</td>\n",
       "      <td>96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>85</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>85</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>82</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>81</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Homozygous A  Homozygous B\n",
       "0            62            96\n",
       "1            62            96\n",
       "2            85            73\n",
       "3            85            73\n",
       "4            82            78\n",
       "5            81            77"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.calculate_allele_frequencies()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Homozygous A</th>\n",
       "      <th>Homozygous B</th>\n",
       "      <th>statistic</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>62</td>\n",
       "      <td>96</td>\n",
       "      <td>7.316456</td>\n",
       "      <td>0.006833</td>\n",
       "      <td>0.020498</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>62</td>\n",
       "      <td>96</td>\n",
       "      <td>7.316456</td>\n",
       "      <td>0.006833</td>\n",
       "      <td>0.020498</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>85</td>\n",
       "      <td>73</td>\n",
       "      <td>0.911392</td>\n",
       "      <td>0.339745</td>\n",
       "      <td>0.509618</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>85</td>\n",
       "      <td>73</td>\n",
       "      <td>0.911392</td>\n",
       "      <td>0.339745</td>\n",
       "      <td>0.509618</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>82</td>\n",
       "      <td>78</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.751830</td>\n",
       "      <td>0.751830</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>81</td>\n",
       "      <td>77</td>\n",
       "      <td>0.101266</td>\n",
       "      <td>0.750316</td>\n",
       "      <td>0.751830</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Homozygous A  Homozygous B  statistic    pvalue    qvalue  significant\n",
       "0            62            96   7.316456  0.006833  0.020498         True\n",
       "1            62            96   7.316456  0.006833  0.020498         True\n",
       "2            85            73   0.911392  0.339745  0.509618        False\n",
       "3            85            73   0.911392  0.339745  0.509618        False\n",
       "4            82            78   0.100000  0.751830  0.751830        False\n",
       "5            81            77   0.101266  0.750316  0.751830        False"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.chisquare()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "test.dump()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
